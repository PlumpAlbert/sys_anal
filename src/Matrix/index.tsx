import * as React from 'react'
import { connect, MapStateToProps, MapDispatchToProps } from 'react-redux'
