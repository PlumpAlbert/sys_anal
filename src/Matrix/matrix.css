.Matrix {
    font-family: "Courier Prime Sans", Consolas, "DejaVu Sans Mono", monospace;
    max-width: 30vw;
    overflow: auto;
}

.matrix-table {
    min-width: 100%;
    empty-cells: show;
}

.matrix-table td:nth-child(even) {
    background-color: hsla(193, 75%, 75%, .6);
    color: hsl(193, 25%, 18%);
}

.matrix-table tr {
    display: flex;
    align-items: center;
}

.matrix-table td,
.matrix-table th {
    width: 1.5rem;
    min-height: 1.2rem;
    flex: 1;
}

.Matrix,
.matrix-table th {
    color: hsl(200, 90%, 70%);
}

.matrix-table td {
    color: hsl(192, 100%, 95%);
}