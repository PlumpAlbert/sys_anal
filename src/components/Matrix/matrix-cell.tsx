import * as React from "react";
import {
  updateLink,
  appendLink,
  removeLink,
  GraphNode,
  Link
} from "../../store";
import { connect, MapDispatchToProps } from "react-redux";
import { implementsInterface } from "../../index";

interface AdjacencyProps {
  sourceNode: GraphNode;
  targetNode: GraphNode;
  linkId?: number;
  initialValue: number;
}

interface IncidenceProps {
  nodeId: number;
  linkId: number;
  initialValue: number;
}

interface IState {
  value: number;
}

type TProps = (AdjacencyProps | IncidenceProps) & IDispatchProps;

function isAdjacencyProps(obj: any): obj is AdjacencyProps {
  const checker = obj as AdjacencyProps;
  if (checker.sourceNode && checker.targetNode) return true;
  return false;
}

export class MatrixCell extends React.PureComponent<TProps, IState> {
  constructor(props: TProps) {
    super(props);
    this.state = { value: props.initialValue };
  }
  adjacencyClick = (event: React.MouseEvent<HTMLTableDataCellElement>) => {
    if (isAdjacencyProps(this.props)) {
      let { value } = this.state;
      let {
        appendLink,
        removeLink,
        linkId,
        sourceNode,
        targetNode
      } = this.props;
      if (value === 1 && linkId !== undefined)
        removeLink(linkId, sourceNode, targetNode);
      else appendLink(sourceNode, targetNode);
    }
    this.setState(({ value }) => ({ value: value === 0 ? 1 : 0 }));
  };
  /**
   * TODO create implementation
   * @param event
   */
  incidenceClick = (event: React.MouseEvent<HTMLTableDataCellElement>) => {
    let { value } = this.state;
    let newValue = 0;
    if (value === 0) newValue = 1;
    else if (value === 1) newValue = -1;
    this.setState({ value: newValue });
  };
  render() {
    if (isAdjacencyProps(this.props)) {
      return <td onClick={this.adjacencyClick}>{this.state.value}</td>;
    } else {
      return <td onClick={this.incidenceClick}>{this.state.value}</td>;
    }
  }
}

interface IDispatchProps {
  appendLink: (
    sourceNode: GraphNode,
    targetNode: GraphNode,
    twoWay?: boolean,
    label?: string
  ) => void;
  updateLink: (index: number, changes: { [p in keyof Link]?: Link[p] }) => void;
  removeLink: (
    index: number,
    sourceNode: GraphNode,
    targetNode: GraphNode
  ) => void;
}

const mapD2P: MapDispatchToProps<
  IDispatchProps,
  AdjacencyProps | IncidenceProps
> = (dispatch, ownProps) => ({
  appendLink: (sourceNode, targetNode, twoWay, label) =>
    dispatch(appendLink({ sourceNode, targetNode, twoWay, label })),
  updateLink: (index, changes) => dispatch(updateLink({ index, changes })),
  removeLink: (index, sourceNode, targetNode) =>
    dispatch(removeLink({ index, sourceNode, targetNode })),
  ...ownProps
});

export default connect(
  null,
  mapD2P
)(MatrixCell);
