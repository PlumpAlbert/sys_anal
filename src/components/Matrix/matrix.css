.matrix {
  font-family: "Google Sans", "FantasqueSansMono Nerd Font", Consolas,
    "DejaVu Sans Mono", monospace;
  overflow: auto;
}

.matrix-content {
  display: inline-block;
}
.matrix-content input {
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  background: transparent;
  border: 0;
  border-top: 2px solid hsl(200, 50%, 60%);
  color: hsl(200, 90%, 70%);
}
.matrix-content input::placeholder {
  color: hsl(200, 90%, 70%);
}
.matrix-content input:focus {
  outline: none;
}

.matrix-wrapper {
  display: flex;
  width: fit-content;
}
.matrix-wrapper td,
.matrix-wrapper th,
.matrix-wrapper button {
  cursor: pointer;
  user-select: none;
  width: 2rem;
  min-height: 1.2rem;
  flex: 1;
}
.matrix-wrapper button {
  background: transparent;
  border: none;
  font-weight: 700;
  color: hsl(200, 90%, 70%);
}

.matrix-table {
  empty-cells: show;
}
.matrix-table td:nth-child(even),
.matrix-table th:nth-child(even) {
  background-color: hsla(193, 75%, 75%, 0.6);
  color: hsl(193, 25%, 18%);
}
.matrix-table > thead > tr {
  border-bottom: 2px solid hsl(200, 50%, 60%);
}
.matrix-table tr {
  display: flex;
  align-items: center;
}

.matrix,
.matrix-table th {
  color: hsl(200, 90%, 70%);
}

.matrix-table td {
  color: hsl(192, 100%, 95%);
}
