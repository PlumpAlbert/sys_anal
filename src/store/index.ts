import * as d3 from "d3";
import { Reducer } from 'redux'
import { createStandardAction } from "typesafe-actions";

export interface GraphNode extends d3.SimulationNodeDatum {
  id: number;
  label?: string;
}

export interface Link extends d3.SimulationLinkDatum<GraphNode> {
  label: string;
  twoWay: boolean;
}

export interface IGlobalState {
  nodes: GraphNode[];
  links: Link[];
}

enum ActionConst {
  updateNode = "@GLOBALSTATE/updateNode",
  updateLink = "@GLOBALSTATE/updateLink"
}

export const updateLink = createStandardAction(ActionConst.updateLink)<{
  link: Link,
  index: number
}>();
export const updateNode = createStandardAction(ActionConst.updateNode)<{
  node: GraphNode,
  index: number
}>();
type TActions = ReturnType<typeof updateLink> | ReturnType<typeof updateNode>;

const initNodes: GraphNode[] = [
  { id: 0 },
  { id: 1 },
  { id: 2 },
  { id: 3 },
  { id: 4 },
  { id: 5 },
  { id: 6 },
  { id: 7 },
  { id: 8 },
  { id: 9 },
  { id: 10 }
];
const initLinks: Link[] = [
  { source: initNodes[0], target: initNodes[1], twoWay: false, label: 'e0' },
  { source: initNodes[2], target: initNodes[1], twoWay: true, label: 'e1' },
  { source: initNodes[4], target: initNodes[0], twoWay: false, label: 'e2' },
  { source: initNodes[2], target: initNodes[5], twoWay: false, label: 'e3' },
  { source: initNodes[4], target: initNodes[6], twoWay: true, label: 'e4' },
  { source: initNodes[3], target: initNodes[7], twoWay: false, label: 'e5' },
  { source: initNodes[7], target: initNodes[5], twoWay: true, label: 'e6' }
];

export const mainReducer: Reducer<IGlobalState, TActions> = (state = {
  nodes: initNodes,
  links: initLinks
}, action) => {
  switch (action.type) {
    case ActionConst.updateNode: {
      let nodes = [...state.nodes];
      nodes.splice(action.payload.index, 1, action.payload.node);
      return { ...state, nodes };
    }
    case ActionConst.updateLink: {
      let links = [...state.links];
      links.splice(action.payload.index, 1, action.payload.link);
      return { ...state, links };
    }
  }
  return state;
};
