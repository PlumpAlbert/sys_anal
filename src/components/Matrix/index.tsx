import * as React from "react";
import { connect, MapDispatchToProps, MapStateToProps } from "react-redux";
import {
  GraphNode,
  IGlobalState,
  Link,
  updateLink,
  updateNode
} from "../../store";
import "./matrix.css";

interface IOwnProps {
  type: "Adjacency" | "Incidence";
}

export type TProps = IOwnProps & IStoreProps & IDispatchProps;

function onBlur(
  event: React.MouseEvent<HTMLButtonElement>,
  data: Link[],
  action: IDispatchProps["appendLinks"]
): void;
function onBlur(
  event: React.FocusEvent<HTMLInputElement>,
  data: GraphNode[],
  action: IDispatchProps["appendNodes"]
): void;
function onBlur(
  event:
    | React.FocusEvent<HTMLInputElement>
    | React.MouseEvent<HTMLButtonElement>,
  data: GraphNode[] | Link[],
  action: IDispatchProps[keyof IDispatchProps]
): void {
  if (implementsInterface<IDispatchProps["appendLinks"]>(action)) {
    return action(data.length);
  }
  if (event.target instanceof HTMLInputElement) {
    let { value } = event.target;
    if (!value) return;
    if (implementsInterface<GraphNode[]>(data)) {
      return action(data.length, {
        id: data[data.length - 1].id + 1,
        label: value
      });
    }
  }
}

export const Matrix: React.FC<TProps> = props => {
  console.log(
    `Matrix (${props.type})`,
    new Date(Date.now()).toLocaleTimeString()
  );
  let table: JSX.Element;
  let columnInput: JSX.Element | null = null;
  if (props.type === "Adjacency") {
    table = (
      <table className="matrix-table">
        <thead>
          <tr>
            <th />
            {props.nodes.map(node => (
              <th key={`head_${node.id}`}>{node.id}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {props.nodes.map(node => {
            let matches: number[] = [];
            for (let i = 0; i < props.links.length; i++) {
              // If link starts from current node
              if (props.links[i].source === node) {
                matches.push((props.links[i].target as GraphNode).id);
              }
              // If link is two-way and current node at the end
              else if (
                props.links[i].target === node &&
                props.links[i].twoWay
              ) {
                matches.push((props.links[i].source as GraphNode).id);
              }
            }
            return (
              <tr key={`row_${node.id}`}>
                <th>{node.id}</th>
                {props.nodes.map(v => {
                  if (matches.includes(v.id))
                    return <td key={`${node.id}_${v.id}`}>1</td>;
                  return <td key={`${node.id}_${v.id}`}>0</td>;
                })}
              </tr>
            );
          })}
        </tbody>
      </table>
    );
  }
  // if (props.type === 'Incidence')
  else {
    table = (
      <table className="matrix-table">
        <thead>
          <tr>
            <th />
            {props.links.map(link => (
              <th key={`head_${link.label}`}>{link.label}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {props.nodes.map(node => {
            let content = props.links.map(link => {
              if (link.source === node) return 1;
              else if (link.target === node) {
                if (link.twoWay) return 1;
                return -1;
              }
              return 0;
            });
            return (
              <tr key={`row_${node.id}`}>
                <th>{node.id}</th>
                {content.map((v, i) => (
                  <td key={`${node.id}_${i}`}>{v}</td>
                ))}
              </tr>
            );
          })}
        </tbody>
      </table>
    );
    columnInput = (
      <button onClick={e => onBlur(e, props.links, props.appendLinks)}>
        +
      </button>
    );
  }
  return (
    <div className="matrix">
      <h2 className="matrix-header">{`${props.type} matrix`}</h2>
      <div className="matrix-content">
        <div className="matrix-wrapper">
          {table}
          {columnInput}
        </div>
        <input
          type="text"
          placeholder="+"
          onBlur={e => onBlur(e, props.nodes, props.appendNodes)}
        />
      </div>
    </div>
  );
};

interface IStoreProps {
  nodes: GraphNode[];
  links: Link[];
}

const mapS2P: MapStateToProps<IStoreProps, IOwnProps, IGlobalState> = (
  { nodes, links },
  ownprops
) => ({
  nodes,
  links,
  ...ownprops
});

interface IDispatchProps {
  appendNodes: (index: number, node: GraphNode) => void;
  appendLinks: (index: number) => void;
}

const mapD2P: MapDispatchToProps<IDispatchProps, IOwnProps> = (
  dispatch,
  { type }
) => ({
  appendNodes: (index, node) =>
    dispatch(
      updateNode({
        index,
        node
      })
    ),
  appendLinks: index =>
    dispatch(
      updateLink({
        index,
        link: {
          label: `e${index}`,
          source: { id: 0 },
          target: { id: 1 },
          twoWay: false
        }
      })
    ),
  type
});

export default connect(
  mapS2P,
  mapD2P
)(Matrix);
