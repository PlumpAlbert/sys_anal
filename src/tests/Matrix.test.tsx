import { shallow } from 'enzyme'
import { Matrix } from '../Matrix';
import { GraphNode, Link } from "../store";
import * as React from "react";

describe('Testing Matrix component', () => {
  let nodes: GraphNode[] = [];
  let links: Link[] = [];
  beforeEach(() => {
    nodes = [
      { id: 1, label: 'A' },
      { id: 1, label: 'B' },
      { id: 1, label: 'C' }
    ];
    links = [
      { source: nodes[0], target: nodes[1], twoWay: false, label: 'e0' },
      { source: nodes[2], target: nodes[1], twoWay: false, label: 'e1' },
      { source: nodes[2], target: nodes[2], twoWay: true, label: 'e2' }
    ]
  });
  it('Creates incidence matrix correctly', () => {
    expect(
      shallow(<Matrix type='Incidence' nodes={nodes} links={links} />).html()
    ).toMatchSnapshot();
  });
  it('Creates adjacency matrix correctly', () => {
    expect(
      shallow(<Matrix type='Adjacency' nodes={nodes} links={links} />).html()
    ).toMatchSnapshot();
  });
});
