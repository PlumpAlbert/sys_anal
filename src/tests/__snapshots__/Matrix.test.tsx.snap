// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Matrix component Creates adjacency matrix correctly 1`] = `"<div class=\\"Matrix\\"><h2 class=\\"Matrix-header\\">Adjacency matrix</h2><table class=\\"matrix-table\\"><thead><tr><th></th><th>1</th><th>2</th><th>3</th></tr></thead><tbody><tr><th>1</th><td>0</td><td>1</td><td>0</td></tr><tr><th>2</th><td>0</td><td>0</td><td>1</td></tr><tr><th>3</th><td>0</td><td>1</td><td>1</td></tr></tbody></table></div>"`;

exports[`Testing Matrix component Creates incidence matrix correctly 1`] = `"<div class=\\"Matrix\\"><h2 class=\\"Matrix-header\\">Incidence matrix</h2><table class=\\"matrix-table\\"><thead><tr><th> </th><th>e0</th><th>e1</th><th>e2</th></tr></thead><tbody><tr><th>1</th><td>1</td><td>0</td><td>0</td></tr><tr><th>2</th><td>-1</td><td>1</td><td>0</td></tr><tr><th>3</th><td>0</td><td>1</td><td>1</td></tr></tbody></table></div>"`;
